python/ray/serve/autoscaling_metrics.py

```diff
@@ -24,39 +28,51 @@ def start_metrics_pusher(
         collection_callback: a callable that returns the metric data points to
           be sent to the the controller. The collection callback should take
           no argument and returns a dictionary of str_key -> float_value.
-        controller_handle: actor handle to Serve controller.
+        metrics_process_func: actor handle function.
+        stop_event: the daemon thread will be closed when the stop event is set
     """
 
     def send_once():
         data = collection_callback()
+
         # TODO(simon): maybe wait for ack or handle controller failure?
-        return controller_handle.record_autoscaling_metrics.remote(
-            data=data, send_timestamp=time.time()
-        )
+        return metrics_process_func.remote(data=data, send_timestamp=time.time())
 
-    def send_forever():
+    def send_forever(stop_event):
         last_ref: Optional[ray.ObjectRef] = None
         last_send_succeeded: bool = True
 
         while True:
             start = time.time()
-
-            if last_ref:
-                ready_refs, _ = ray.wait([last_ref], timeout=0)
-                last_send_succeeded = len(ready_refs) == 1
-            if last_send_succeeded:
-                last_ref = send_once()
+            if stop_event and stop_event.is_set():
+                return
+
+            if ray.is_initialized():
```

Comment:
why is this necessary?

Reply:
If ray is not initiated (I randomly hit the issue when doing unit tests), the background thread will hit worker connect issue. 
This change is for putting some efforts to make sure ray is always initiated before sending metrics, although it is not always guaranteed.

Reply:
ideally this shouldn't happen (user can't get a handle before ray isn't initialized).. this is probably due to the fact that we shutdown ray between tests but the thread remains?

Reply:
+1, this is probably an issue on the testing side. If possible, it would be nice to remove this if-condition since it doesn't seem necessary. If we can't, we should add a comment explaining that it's there to satisfy the unit tests.